### PR TITLE
fix(ce-code-review,ce-doc-review): make cross-model peers survive the tool-call ceiling

### PR DESCRIPTION
**Pull request by tmchow**
## Summary

The cross-model review passes in `ce-doc-review` and `ce-code-review` could silently lose their independent peer. Both skills told the orchestrator to launch the peer CLI and hold a single shell tool call open for its full 2-5 minute runtime; on harnesses that cap a tool call at ~2 minutes, the harness killed the supervising shell mid-run, and because the pass is deliberately non-blocking, the result was indistinguishable from "no peer was available" — a review that quietly dropped its second opinion with no signal.

Peers now run as **detached, supervised jobs** the orchestrator polls in short calls, so no tool call ever spans a peer's runtime. The pass survives any tool-call ceiling on every host. Separately, reviews now record **which model actually served** each peer run, instead of asserting the requested model as fact.

Related: #878, #1115

## What changed

A new bundled job runner (`peer-job-runner.py`, byte-duplicated into both skills with a parity test) owns the lifecycle: `start` detaches the existing peer script into its own session and returns a job id in under a second; a supervisor inside the detached process watches liveness and enforces caps; `wait`/`status`/`reap`/`result` are all bounded. The skills' launch contract changed from "await one long call" to "start → bounded-wait → reap → fold in what finished by the aggregate deadline," and a started peer that ends in any non-`done` state is now named in Coverage rather than vanishing.

The worker scripts additionally parse the served-model receipt where the CLI reports one (the `claude` route's `modelUsage`) and label every other route `unverified` — the cross-model agreement bonus is kept, but the report no longer claims a specific peer model it cannot prove.

## How dispatch and reaping work

The whole design turns one long-lived tool call into many short ones. `start` detaches the peer into its own session and returns immediately; the orchestrator then polls that job in bounded calls between its other work, and `reap` guarantees a terminal outcome at the aggregate deadline — nothing is ever left running or silently lost.

```mermaid
sequenceDiagram
    autonumber
    participant O as Orchestrator (review skill)
    participant R as peer-job-runner.py
    participant S as Supervisor (detached session)
    participant W as Peer CLI (read-only)

    Note over O,W: start -- one short call, returns a job id in under ~2s
    O->>R: start --skill ... -- (peer command)
    R->>R: mint durable job dir (0700, in /tmp)
    R->>S: setsid double-fork (survives the tool-call kill)
    S->>W: exec peer CLI against the work tree
    R-->>O: job id
    Note over S,W: peer runs 2-5 min in its own session -- no tool call spans it
    W-->>S: heartbeat keeps out.log growing (liveness)
    W->>W: on success, publish reviewer-provider.json

    loop until terminal OR 610s since start
        O->>R: wait --max-secs 30 job-id
        R-->>O: state (running / done / timeout / ...)
        Note over O: do other Stage 4/5 work between polls
    end

    opt still running at the deadline
        O->>R: reap job-id
        R->>S: signal supervisor (it owns TERM -> grace -> KILL + classify)
        Note over R,S: supervisor gone? the runner sweeps the worker pgid itself
        O->>R: wait --max-secs 10 (reap is async)
    end

    O->>R: result --path artifact (fd-ownership-checked read)
    R-->>O: findings JSON, or exit 4 -> treat as no file
    O->>O: fold in / classify skip reason / delete job dir
```

The subtle part is `reap`. When the deadline hits, who cleans up depends on whether the supervisor is still alive — and a peer child can outlive its own session leader, so the sweep must target the process **group**, not just the leader pid (this is the class of bug the review pass kept catching in its own runner):

```mermaid
flowchart TD
    A["reap job-id"] --> B{"already terminal?"}
    B -- yes --> Z["done: existing record stands"]
    B -- no --> C{"supervisor alive?"}
    C -- yes --> D["signal supervisor: it runs<br/>TERM -> grace -> KILL and writes<br/>the terminal record"]
    D --> Z
    C -- "no (gone)" --> E{"have a worker pid?"}
    E -- no --> H["write died-without-result"]
    E -- yes --> F["sweep the worker process GROUP<br/>kill_tree(pgid) -- even if the leader<br/>already exited, a child can survive in it"]
    F --> G{"leader was alive?"}
    G -- yes --> I["classify timeout (reaped a live tree)"]
    G -- no --> H
    H --> Z
    I --> Z
```

## Design decisions worth a look

- **Python, not bash, for the runner.** It needs `os.setsid` (macOS ships no `setsid` binary), fd-level ownership checks, and five subcommands with real error paths — the profile the repo's own "prefer Python for pipeline scripts" learning routes to Python.
- **Detach via a new session, because `nohup` isn't enough.** Measured across three hosts: plain `nohup` is reaped the instant the launching tool call ends on Codex and Cursor; only a new session (setsid double-fork) survives everywhere. This is the whole reason the lifecycle exists and is documented in `docs/solutions/`.
- **Label-only model identity.** Agreement promotion still fires on routes without a served-model receipt; the report marks the peer model unverified rather than gating the bonus off (which would disable it on three of four routes today). Invalid model ids fail loudly on all tested CLIs, so the residual is narrow.
- **Job state is hardened for shared `/tmp`.** Every read is fd-ownership-checked before content reaches agent context; directories/files are created 0700/0600 with no-follow and atomic rename; consumed job dirs are deleted after fold-in.

## Validation

- Full `bun test` (2049 pass), `release:validate`, and `plugin:validate` all green. New regression tests cover the runner lifecycle, receipt normalization, byte-parity of both duplicated assets, and the specific edge cases review surfaced.
- **Live-verified with real peers on two hosts.** On Claude Code, a real Codex peer ran end-to-end through the runner — detached, stayed alive ~6 minutes past the old 240s reap point via the liveness heartbeat, reached `done`, and folded in with the receipt fields populated. On `cursor-agent` (loaded from this worktree via `--plugin-dir`), a full `/ce-code-review` run followed the new start-poll-reap contract with "no ambiguities, no errors."
- The cross-model adversarial peer earned its keep during that verification: it flagged a real busy-loop in the heartbeat code (a `0`-interval `sleep`), which is fixed in the final commit here.

The one detach behavior not yet exercised is default-sandboxed Codex, covered by the skill-creator eval; a failure there is designed to surface as a stop condition, not a silent ship.

## New concepts

**Detaching work to outlive a harness tool-call ceiling.** Agent harnesses cap how long a single shell tool call may run. Any skill that shells out to slow work (a peer model CLI, a build, a long analysis) and *awaits it in one call* dies when the harness kills that call — and if the work was non-blocking, the failure is invisible.

The fix is to never let one tool call span the work's runtime: `start` mints a durable job directory, detaches the worker into its own session, and returns immediately; the orchestrator then polls that job's state in sub-second calls between its other work.

```bash
# Plain nohup is NOT enough — measured: Codex and Cursor reap it when the
# launching tool call ends. A new session (setsid) survives; macOS has no
# setsid binary, so a perl/POSIX double-fork does it:
perl -e 'use POSIX qw(setsid); exit if fork; setsid(); exit if fork;
  open(STDIN,"<","/dev/null"); open(STDOUT,">","/dev/null");
  exec "/bin/bash","-c",$ARGV[0];' "$CMD"
```

Reach for this whenever delegated work can outlive the host's tool-call budget. Skip it for work that reliably finishes in seconds — the single-call contract is simpler and fine there.

---

[Compound Engineering](https://github.com/EveryInc/compound-engineering-plugin)
Claude Code

### CODE REVIEW

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+        pid_doc = None
+    sup_pid = pid_doc.get("supervisor_pid") if isinstance(pid_doc, dict) else None
+    sup_pgid = pid_doc.get("supervisor_pgid") if isinstance(pid_doc, dict) else None
+    worker_pid = pid_doc.get("worker_pid") if isinstance(pid_doc, dict) else None
+
+    if isinstance(sup_pid, int) and _pid_alive(sup_pid):
+        # The supervisor owns TERM-grace-KILL and the terminal classification;
+        # signal it and return immediately.
+        if (isinstance(sup_pgid, int) and _killpg_quiet(sup_pgid, signal.SIGTERM)) \
+                or _kill_quiet(sup_pid, signal.SIGTERM):
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py:869]: > When the detached supervisor has already been killed or crashed but remains as a zombie, `os.kill(pid, 0)` still succeeds, so this branch sends TERM and returns without killing `worker_pid` or writing a terminal record.

Addressed in f78dd5eb. Good catch — `kill -0` is true for a zombie, so the live-supervisor branch of `cmd_reap` no longer returns immediately after signaling. It now polls `job_state` for a terminal record on a bounded window (`min(grace, 1.0)`s); if the supervisor classifies within it we return, otherwise we fall through to the existing self-cleanup path that runs `kill_tree(worker_pid, ...)` and writes the terminal record itself — so a zombie/wedged supervisor can no longer leave the job stuck `running`. Reap stays under its ~1s budget.

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 2]
```diff
@@ -0,0 +1,953 @@
...
+        for descendant in _descendants_deepest_first(pid):
+            _kill_quiet(descendant, sig)
+        _kill_quiet(pid, sig)
+
+
+def kill_tree(root_pid: int, grace: float) -> bool:
+    """TERM the pid's process group (workers are started as group leaders),
+    falling back to a deepest-first tree walk; grace, then KILL survivors."""
+    if not _pid_alive(root_pid):
+        return False
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py]: > If the leader has already exited but a child is still running in the same process group, this early-return skips the `killpg` sweep entirely.

Fixed in `75f9cdcd`. `kill_tree` no longer early-returns on a dead leader: it records `leader_alive = _pid_alive(root_pid)` up front, then **always** runs the full sweep (snapshot descendants → `SIGTERM` group → grace → `killpg(SIGKILL)` → deepest-first tree-walk fallback → kill root), returning `leader_alive`. The pgid persists while any group member lives even after the leader exits, so the surviving child is now swept instead of orphaned. Only the grace early-break is gated on `leader_alive`, so the alive-leader path is behavior-identical. Byte-parity kept across both skills; `cmp` and `py_compile` clean.

File: tests/skills/peer-job-runner.test.ts [hunk 3]
```diff
@@ -0,0 +1,445 @@
...
+  return dir
+}
+
+// Every supervisor/worker pid the suite detached; teardown proves no orphans
+// (the Verification gate: reaping kills whole trees).
+const trackedPids: number[] = []
+
+function pidAlive(pid: number): boolean {
+  return spawnSync("kill", ["-0", String(pid)]).status === 0
+}
```
[the remark this answers is not in the document]
Comment (tmchow) [reply tests/skills/peer-job-runner.test.ts:45]: > `kill -0` succeeds for a `<defunct>` zombie, so the afterAll orphan assertion could flag a zombie as an orphan and flake.

Fixed in `75f9cdcd`. `pidAlive` now treats a zombie as dead: after `kill -0` succeeds it reads `ps -o state= -p <pid>` and returns false when the state begins with `Z` (macOS/Linux), falling back to the `kill -0` result if `ps` is unavailable. In fairness the orphan assertion stayed green across 10+ repeats here, so this is latent-flake prevention rather than a reproduced failure — but the `kill -0`-counts-zombies-as-alive point is correct and now hardened.

**Comment by tmchow** [comment]
Addressed codex's P2 ("Resolve real interpreters for route-test sandboxes") in `e465a95c`.

The route-test sandbox builds a restricted PATH that deliberately excludes version managers, then symlinked the `command -v python3` result into it. On a pyenv/mise/perlbrew host that result is a wrapper **shim**, which can't exec once the manager is off PATH — so the codex JSON-recovery tests failed with no output artifact (your repro: `command -v python3` = `/root/.pyenv/shims/python3`).

Fix: resolve interpreters (`python3`, `perl`) to their real standalone binary by asking the interpreter itself (`sys.executable` / `$Config{perlpath}`) before linking, so the sandbox links the executable rather than the shim. It's a no-op where the tool already resolves to a real binary, and verified by following a synthetic wrapper shim to its real target. Applied to both route-test suites.

**Comment by tmchow** [comment]
Addressed codex's P2 ("Require findings-shaped output before suppressing fallback") in `08110767` (on the rebased branch).

`out_missing_or_invalid` in the doc worker accepted **any** valid JSON, so a grok response that is valid JSON but not a reviewer return — an error/envelope object with no top-level `findings` array, e.g. a 402 usage-exhausted body — read as usable. That suppressed the `grok-cursor` fallback in `run_provider`, and normalize later dropped the non-findings JSON, so the doc-review cross-model pass could produce no fold-in even though the fallback route would have succeeded.

Fix: require a top-level `findings` array, matching the adversarial (ce-code-review) twin whose check already did exactly this — this was a divergence between the two workers. Pinned the gate in the contract test for both workers so they can't drift again.

(Branch was also rebased onto latest `origin/main`.)

**Comment by tmchow** [comment]
@cursoragent review

**Comment by tmchow** [comment]
Good catch — fixed in `d065c7ac`.

You're right that this is the call-site counterpart to the earlier `kill_tree` hardening: `cmd_reap` only invoked the dead-leader-safe `kill_tree` under `if _pid_alive(worker_pid)`, so a child surviving in the worker's process group after the leader exited was never swept — reap wrote `died-without-result` and left it running. The guard re-defeated the very path `kill_tree` was made safe for.

Fix: call `kill_tree(worker_pid, ...)` whenever `worker_pid` is an int, and classify off its return value (`leader_alive` -> `timeout`, else `died-without-result`). The classification is unchanged; the only behavioral change is that the pgid is now swept in the dead-leader case (the `killpg` targets the pgid, which persists while any group member lives).

Added the regression test you asked for — dead supervisor + dead leader + live child in the worker pgid — verified to **fail** on the guarded call (child survives reap; teardown reports a dangling process) and pass once the sweep is unconditional. Byte-parity kept across both peer-job-runner copies.

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 4]
```diff
@@ -0,0 +1,955 @@
...
+            while time.monotonic() < deadline:
+                if job_state(job_dir) in TERMINAL_STATES:
+                    return 0
+                time.sleep(0.05)
+
+    # Supervisor gone: perform the tree kill and classification ourselves,
+    # with a short grace so reap still returns quickly.
+    killed = False
+    if isinstance(worker_pid, int) and _pid_alive(worker_pid):
+        kill_tree(worker_pid, min(conf["grace"], 1.0))
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py]: > When the supervisor is gone and the worker session leader has already exited/reaped while a child remains in the same pgid, `_pid_alive` is false, so `reap` writes `died-without-result` without signaling that process group.

Already addressed in `d065c7ac` (Cursor Bugbot raised the same call site in parallel). `cmd_reap` now calls `kill_tree(worker_pid, ...)` for any integer `worker_pid` and classifies off its return (`leader_alive` -> `timeout`, else `died-without-result`), so the pgid is swept even when the leader has exited. Both byte-parity copies are fixed, and there is a new regression test (dead supervisor + dead leader + live `sleep` child in the pgid) verified to fail on the guarded call and pass with the unconditional sweep.

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 5]
```diff
@@ -0,0 +1,959 @@
...
+        return "running"
+    return "never-started"
+
+
+# --- process-tree control -----------------------------------------------------
+
+def _pid_alive(pid: int) -> bool:
+    try:
+        os.kill(pid, 0)
+        return True
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py:326]: > the runtime helper still treats a zombie worker leader as alive: in the dead-supervisor/dead-leader scenario, `kill_tree()` returns `true`, so `reap` records `timeout` instead of `died-without-result`.

Fixed in `456395c7` — and this also explains why the new sweep test passed for me but failed for you: `kill_tree` derived `leader_alive` from `_pid_alive` (`os.kill(pid, 0)`), which succeeds for a `<defunct>` zombie, so whether reap saw `timeout` or `died-without-result` depended on whether it caught the brief zombie window before init reaped the orphaned leader.

Added a zombie-aware `_pid_running` (`kill -0`, then a `ps -o state=` `Z` check — the same signal your TS helper uses) and classify the reap off it. To make it a real guard rather than a timing coin-flip, added a **deterministic** zombie unit test to the python fixture: fork a child that exits and leave it unreaped, then assert `kill -0` still sees it but `_pid_running` does not. Byte-parity kept across both runner copies.

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 6]
```diff
@@ -0,0 +1,983 @@
...
+        if now - start_t >= conf["hard"]:
+            _reap_worker(proc, conf)
+            state, reason = "timeout", f"hard cap {conf['hard']:g}s exceeded"
+            break
+        _interruptible_sleep(conf["poll"], flag)
+    # An externally killed worker can leave group members behind (its shell's
+    # children); sweep the group before publishing so no orphan outlives the
+    # terminal record. A pgid cannot be recycled while members remain.
+    _killpg_quiet(proc.pid, signal.SIGTERM)
+    _killpg_quiet(proc.pid, signal.SIGKILL)
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py:587]: > The bundled cross-model scripts start provider calls under job control (`set -m`), so those provider processes live in a separate pgid (`ACTIVE_PEER_PID`), not `proc.pid` ... reproduced a `set -m; sleep 60 & exit 0` worker being reported `done` while the child stayed alive.

Fixed in `ef91a099`. You are right that the provider runs in its own pgid — that is deliberate (`set -m` lets the worker `killpg` the provider without killing itself, since the worker shares the runner`s group). Tracing the runner: its **reap** paths (`_reap_worker`, and `cmd_reap` when the supervisor is gone) already sweep separate-pgid children, because they snapshot the worker`s full descendant tree by ppid and individually `SIGKILL` it — the provider is a child of the worker regardless of pgid. The gap you reproduced is the **clean-exit (`done`)** path: the runner never sweeps there (it trusts the worker), and the supervisor`s final sweep is `killpg(worker_pgid)`-only, which misses the provider group. The runner cannot close this itself — after a clean exit the orphan reparents to init, so a descendant walk finds nothing.

So the fix is at the worker, where the provider pgid (`ACTIVE_PEER_PID`) is known: `run_codex_cmd` and `run_timeout_cmd` now `reap "$pid"` (group-reap) after `wait`, on every completion path, not just idle/hard/trap. `reap()` returns immediately when the group is already empty, so it is free in the common case. A contract guard pins the sweep in both workers. Applied to both byte-parallel scripts.

File: skills/ce-code-review/scripts/cross-model-adversarial-review.sh [hunk 7]
```diff
@@ -455,6 +554,15 @@ run_provider() {
...
     log "provider $provider produced no usable schema-shaped output; skipping fold-in"
+    # Surface a bounded tail of the peer's raw output so the orchestrator can
+    # reason about WHY it was skipped (quota/usage-limit exhaustion vs an ordinary
+    # empty review) and, in a repeated-pass session, deprioritize an exhausted
+    # route. Harness-agnostic: the agent classifies from the text; this only makes
+    # the evidence visible in out.log. Bash builtins only (the route sandbox has no
+    # tail/tr), and PEERLOG is a small error body on a failed route.
+    if [ -s "$PEERLOG" ]; then
+      _pt="$(< "$PEERLOG")"; _pt="${_pt//$'\n'/ }"; log "  peer skip evidence: ${_pt: -300}"
+    fi
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-code-review/scripts/cross-model-adversarial-review.sh:571]: > `run_timeout_cmd` still launches those routes with stderr redirected to `/dev/null` instead of `$PEERLOG` ... capture stderr into the bounded peer log before tailing it here.

Fixed in `98429cb4`. Correct — the `peer skip evidence` surface I added only tailed PEERLOG (stdout), and `run_timeout_cmd` (claude/grok/cursor-agent) dropped stderr, so an auth/quota message on stderr produced no evidence on exactly those routes.

One refinement on the suggested fix: I did **not** merge stderr into PEERLOG, because PEERLOG must stay clean stdout — the findings brace-match and the receipt `jq`-parse (`.modelUsage`) both read it, and stderr noise would corrupt the receipt for the claude route. Instead stderr goes to a separate `PEERERR` file, reset per attempt, and the skip-evidence block now tails **both** (stdout for grok`s 402, stderr for claude/cursor auth/quota). A contract guard pins the separate-stderr capture so it can`t regress to `/dev/null`. Applied to both byte-parallel workers.

File: skills/ce-doc-review/scripts/peer-job-runner.py [hunk 8]
```diff
@@ -0,0 +1,983 @@
...
+    if isinstance(worker_pid, int):
+        worker_leader_alive = kill_tree(worker_pid, min(conf["grace"], 1.0))
+    if worker_leader_alive:
+        word, reason = "timeout", (
+            "reaped by request; supervisor was gone, worker tree killed by reap"
+        )
+    else:
+        word, reason = "died-without-result", (
+            "supervisor and worker both gone without a terminal record"
+        )
```
[the remark this answers is not in the document]
Comment (tmchow) [reply skills/ce-doc-review/scripts/peer-job-runner.py:921]: > This fallback branch never rereads `meta.json`/`result_path`; once the leader is gone it records `died-without-result`, so `result <job>` exits 3 even though the artifact exists.

Good find, and a real correctness bug rather than just cleanup — fixed in `3553e6b3`. When the supervisor dies mid-run and the worker then publishes its declared result and exits before reap runs, the fallback saw a dead leader and recorded `died-without-result`, discarding a review that actually succeeded.

`cmd_reap` now reads the declared `result_path` from `meta.json` and classifies from the artifact (`done`, or `failed` if it exceeds the byte cap) before falling back to timeout / died-without-result — mirroring `classify_exit`, so the two agree. Added the regression test from your repro: kill the supervisor, let the stub publish its result and exit, reap — verified to fail on the old code (`died-without-result`) and pass now (`done`, and `result` emits the bytes). Byte-parity kept across both runner copies.
